src/PowerShellEditorServices/Language/TokenOperations.cs

```diff
@@ -21,103 +21,120 @@ internal static class TokenOperations
-                MatchTokenElements(tokens, s_openingParens, TokenKind.RParen, RegionKindNone)
-            );
-
-            // Find contiguous here strings @' -> '@
-            foldableRegions.AddRange(
-                MatchTokenElement(tokens, TokenKind.HereStringLiteral, RegionKindNone)
-            );
-
-            // Find unopinionated variable names ${ \n \n }
-            foldableRegions.AddRange(
-                MatchTokenElement(tokens, TokenKind.Variable, RegionKindNone)
-            );
-
-            // Find contiguous here strings @" -> "@
-            foldableRegions.AddRange(
-                MatchTokenElement(tokens, TokenKind.HereStringExpandable, RegionKindNone)
-            );
+            Stack<Token> tokenCurlyStack = new Stack<Token>();
+            Stack<Token> tokenParenStack = new Stack<Token>();
+            foreach (Token token in tokens)
+            {
+                // Find matching braces  { -> }
+                // Find matching hashes @{ -> }
+                if ((token.Kind == TokenKind.LCurly) || (token.Kind == TokenKind.AtCurly)) {
+                    tokenCurlyStack.Push(token);
+                }
+                if ((tokenCurlyStack.Count > 0) && (token.Kind == TokenKind.RCurly)) {
+                    refList.SafeAdd(CreateFoldingReference(tokenCurlyStack.Pop(), token, RegionKindNone));
+                }
+                // Find matching parentheses     ( -> )
+                // Find matching array literals @( -> )
+                // Find matching subexpressions $( -> )
+                if ((token.Kind == TokenKind.LParen) || (token.Kind == TokenKind.AtParen) || (token.Kind == TokenKind.DollarParen)) {
+                    tokenParenStack.Push(token);
+                }
+                if ((tokenParenStack.Count > 0) && (token.Kind == TokenKind.RParen)) {
+                    refList.SafeAdd(CreateFoldingReference(tokenParenStack.Pop(), token, RegionKindNone));
+                }
+                // Find contiguous here strings @' -> '@
+                // Find unopinionated variable names ${ \n \n }
+                // Find contiguous expandable here strings @" -> "@
+                if ((
+                    (token.Kind == TokenKind.HereStringLiteral) ||
+                    (token.Kind == TokenKind.Variable) ||
+                    (token.Kind == TokenKind.HereStringExpandable)
+                    ) && (token.Extent.StartLineNumber != token.Extent.EndLineNumber)) {
+                    refList.SafeAdd(CreateFoldingReference(token, token, RegionKindNone));
+                }
+            }
 
             // Find matching comment regions   #region -> #endregion
-            foldableRegions.AddRange(
-                MatchCustomCommentRegionTokenElements(tokens, RegionKindRegion)
-            );
-
+            // Given a list of tokens, find the tokens that are comments and
+            // the comment text is either `# region` or `# endregion`, and then use a stack to determine
```

Comment:
is it `# region` or `#region`? I think the latter, right?

Reply:
Regex is `"^\s*#region\b"` so yes latter.  Fixed.